testing/unittests/jthreadtests.cpp

```diff
@@ -0,0 +1,591 @@
+        pool->start(&lifespan2seconds, "Thread6");
+
+        auto afterDropEndTime = std::chrono::high_resolution_clock::now();
+        auto afterDropDuration = std::chrono::duration_cast<std::chrono::milliseconds>(afterDropEndTime - afterDropStartTime);
+
+        // Should start immediately without throttling
+        CPPUNIT_ASSERT(afterDropDuration.count() < 100);
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        CPPUNIT_ASSERT_EQUAL(2U, pool->runningCount());
+
+        // Test: Pool was above, fell below, and is now at defaultmax again
+        pool->start(&lifespan2seconds, "Thread7"); // Now at capacity (3 threads)
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        CPPUNIT_ASSERT_EQUAL(3U, pool->runningCount());
+
+        auto reThrottleStartTime = std::chrono::high_resolution_clock::now();
+        try
+        {
+            pool->start(&lifespan2seconds, "Thread8", 200); // timeout after 200ms
+            CPPUNIT_FAIL("Start should have thrown an exception as no slots would be available when start delay specified");
+        }
+        catch (IException *e)
+        {
+            auto reThrottleEndTime = std::chrono::high_resolution_clock::now();
+            auto reThrottleDuration = std::chrono::duration_cast<std::chrono::milliseconds>(reThrottleEndTime - reThrottleStartTime);
+
+            CPPUNIT_ASSERT(reThrottleDuration.count() >= 150 && reThrottleDuration.count() <= 300);
+            e->Release();
+        }
+
+        // Test: waitAvailable functionality
+        CPPUNIT_ASSERT(!pool->waitAvailable(100)); // Should timeout since pool is full
+
+        // Wait for threads to finish
+        std::this_thread::sleep_for(std::chrono::seconds(3));
+        CPPUNIT_ASSERT_EQUAL(0U, pool->runningCount());
+
+        // Now waitAvailable should succeed immediately
+        auto waitStartTime = std::chrono::high_resolution_clock::now();
+        CPPUNIT_ASSERT(pool->waitAvailable(1000));
+        auto waitEndTime = std::chrono::high_resolution_clock::now();
+        auto waitDuration = std::chrono::duration_cast<std::chrono::milliseconds>(waitEndTime - waitStartTime);
+
+        CPPUNIT_ASSERT(waitDuration.count() < 50);
+    }
+
+    void testThrottledThreadPoolWithFastThreadCompletion()
+    {
+        // Test the behavior of a throttled thread pool under high load with fast-completing threads
+        pool.setown(createThreadPool("ThrottledTestPoolWithFastThreadCompletion", factory, true, nullptr, 100, 1));
+
+        for (int i = 1; i <= 10; i++)
+        {
+            threadStartedCount = 0;
+            threadCompletedCount = 0;
+
+            for (int i = 1; i <= 200; i++)
+            {
+                StringBuffer threadName;
+                threadName.appendf("Thread%d", i);
+                pool->start(&lifespan1second, threadName.str());
+            }
+
+            std::this_thread::sleep_for(std::chrono::seconds(2));
```

Comment:
would be good to add a comment why expected to be <2 secs (100 immediate, 100 delayed for.., total time expected to be ~..)